From 27b0ff5ccd66eaaeda2ac2a013815c4d34cf9ec9 Mon Sep 17 00:00:00 2001
From: Derek Higgins <higginsd@gmail.com>
Date: Fri, 4 Nov 2011 00:25:34 +0000
Subject: [PATCH 54/95] Undefine libvirt saved instances

Fixes bug 814561

Adding a call to managedSaveRemove if the instance has a
saved instance, so they are now undefined in addition to running
instances during destroy
With test case

Also added myself to Authors

(cherry picked from commit ad7fcf225e126d2a719c04019c4daa1616d2159e)

Change-Id: I760a15d2ab135d7c3d638ca3c4358d8600582411
---
 Authors                         |    1 +
 nova/tests/test_libvirt.py      |   22 ++++++++++++++++++++++
 nova/virt/libvirt/connection.py |   11 +++++++++++
 3 files changed, 34 insertions(+), 0 deletions(-)

diff --git a/Authors b/Authors
index cedd199..f8aa25e 100644
--- a/Authors
+++ b/Authors
@@ -31,6 +31,7 @@ Dave Walker <DaveWalker@ubuntu.com>
 David Pravec <David.Pravec@danix.org>
 David Subiros <david.perez5@hp.com>
 Dean Troyer <dtroyer@gmail.com>
+Derek Higgins <higginsd@gmail.com>
 Devendra Modium <dmodium@isi.edu>
 Devin Carlen <devin.carlen@gmail.com>
 Donal Lafferty <donal.lafferty@citrix.com>
diff --git a/nova/tests/test_libvirt.py b/nova/tests/test_libvirt.py
index 0673af2..f59b2f7 100644
--- a/nova/tests/test_libvirt.py
+++ b/nova/tests/test_libvirt.py
@@ -987,6 +987,28 @@ class LibvirtConnTestCase(test.TestCase):
         _assert_volume_in_mapping('sdg', False)
         _assert_volume_in_mapping('sdh1', False)
 
+    def test_destroy_saved(self):
+        """Ensure destroy calls managedSaveRemove for saved instance"""
+        # Skip if non-libvirt environment
+        if not self.lazy_load_library_exists():
+            return
+
+        mock = self.mox.CreateMock(libvirt.virDomain)
+        mock.destroy()
+        mock.hasManagedSaveImage(0).AndReturn(1)
+        mock.managedSaveRemove(0)
+        mock.undefine()
+
+        self.mox.ReplayAll()
+
+        def fake_lookup_by_name(instance_name):
+            return mock
+
+        conn = connection.LibvirtConnection(False)
+        self.stubs.Set(conn, '_lookup_by_name', fake_lookup_by_name)
+        instance = {"name": "instancename", "id": "instanceid"}
+        conn.destroy(instance, [])
+
 
 class NWFilterFakes:
     def __init__(self):
diff --git a/nova/virt/libvirt/connection.py b/nova/virt/libvirt/connection.py
index 67f4434..ba1dc86 100644
--- a/nova/virt/libvirt/connection.py
+++ b/nova/virt/libvirt/connection.py
@@ -293,6 +293,17 @@ class LibvirtConnection(driver.ComputeDriver):
                     raise
 
             try:
+                # NOTE(derekh): we can switch to undefineFlags and
+                # VIR_DOMAIN_UNDEFINE_MANAGED_SAVE once we require 0.9.4
+                if virt_dom.hasManagedSaveImage(0):
+                    virt_dom.managedSaveRemove(0)
+            except libvirt.libvirtError as e:
+                errcode = e.get_error_code()
+                LOG.warning(_("Error from libvirt during saved instance "
+                              "removal %(instance_name)s. Code=%(errcode)s"
+                              " Error=%(e)s") % locals())
+
+            try:
                 # NOTE(justinsb): We remove the domain definition. We probably
                 # would do better to keep it if cleanup=False (e.g. volumes?)
                 # (e.g. #2 - not losing machines on failure)
-- 
1.7.6.5

